cpp/dolfin/io/cells.cpp

```diff
@@ -0,0 +1,323 @@
+      for (int i = 0; i < num_nodes; ++i)
+        perm[reversed[i]] = i;
+      return perm;
+    }
+  default:
+    throw std::runtime_error("Simplicies can be expressed as TensorProduct");
+  }
+}
+//-----------------------------------------------------------------------------
+std::vector<std::uint8_t> io::cells::lex_to_tp(mesh::CellType type,
+                                               int num_nodes)
+{
+  switch (type)
+  {
+  case mesh::CellType::quadrilateral:
+  {
+    assert((std::sqrt(num_nodes) - std::floor(std::sqrt(num_nodes))) == 0);
+    // Number of nodes in each direction
+    const int n = sqrt(num_nodes);
+
+    std::vector<std::uint8_t> permutation(num_nodes);
+    int j = 0;
+    std::vector<std::uint8_t> rows(n);
+    std::iota(std::next(rows.begin()), std::prev(rows.end()), 2);
+    rows.front() = 0;
+    rows.back() = 1;
+
+    std::vector<std::uint8_t>::iterator row;
+    for (row = rows.begin(); row != rows.end(); ++row)
+    {
+      permutation[j] = *row;
+      permutation[j + n - 1] = n + *row;
+      j++;
+      for (int index = 0; index < n - 2; ++index)
+      {
+        permutation[j] = (2 + index) * n + *row;
+        j++;
+      }
+      j++;
+    }
+    return permutation;
+  }
+  case mesh::CellType::hexahedron:
+    switch (num_nodes)
+    {
+    case 8:
+      return {0, 1, 3, 2, 4, 5, 7, 6};
+    default:
+      throw std::runtime_error("Higher order hexahedron not supported");
+    }
+  default:
+    throw std::runtime_error("Simplicies can be expressed as TensorProduct");
+  }
+}
+//-----------------------------------------------------------------------------
+std::vector<std::uint8_t> io::cells::vtk_to_dolfin(mesh::CellType type,
+                                                   int num_nodes)
+{
+  switch (type)
+  {
+    {
+    case mesh::CellType::point:
+      return {0};
+    case mesh::CellType::interval:
+      return {0, 1};
+    case mesh::CellType::triangle:
+    {
+      std::vector<std::uint8_t> reversed
+          = io::cells::dolfin_to_vtk(type, num_nodes);
+      std::vector<std::uint8_t> perm(num_nodes);
+      for (int i = 0; i < num_nodes; ++i)
+        perm[reversed[i]] = i;
+      return perm;
+    }
+    case mesh::CellType::tetrahedron:
+      switch (num_nodes)
+      {
+      case 4:
+        return {0, 1, 2, 3};
+      default:
+        throw std::runtime_error("Higher order tetrahedron not supported");
+      }
+    case mesh::CellType::quadrilateral:
+      return io::cells::vtk_to_tp(type, num_nodes);
+    case mesh::CellType::hexahedron:
+      return io::cells::vtk_to_tp(type, num_nodes);
+    default:
+      throw std::runtime_error("Unknown cell type.");
+    }
+  }
+}
+//-----------------------------------------------------------------------------
+Eigen::Array<std::int64_t, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
+io::cells::gmsh_to_dolfin_ordering(
+    Eigen::Array<std::int64_t, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
```

Comment:
Should this be a const reference? e.g. `const Eigen::Ref<const Eigen::Array<...>>&` to avoid a copy?